File: skills/experiment-workspace/scripts/collect-input.py
#!/usr/bin/env python3
"""
ROLE: CUSTOMIZE — implement fetch_metric_summary() below before running.

Input collector — query your data source and produce input.json for analyze-bayesian.py.

Usage:
    python .featbit-release-decision/scripts/collect-input.py <experiment-slug>

Reads:
    .featbit-release-decision/experiments/<slug>/definition.md

Writes:
    .featbit-release-decision/experiments/<slug>/input.json

---
CUSTOMIZATION

Implement the one function below:

    fetch_metric_summary(flag_key, variant, metric, start, end) -> (n, k)

See references/data-source-guide.md for copy-paste patterns:
  §FeatBit  — call the experiment results API
  §Database — run a SQL aggregation query
  §Custom   — call your own metrics service
---
"""

import json
import re
import sys
from pathlib import Path


# ── THE ONE FUNCTION YOU IMPLEMENT ────────────────────────────────────────────

def fetch_metric_summary(
    flag_key: str,
    variant: str,
    metric: str,
    start: str,
    end: str,
) -> tuple[int, int]:
    """
    Return (n_exposed, n_converted) for one variant × metric combination.

    Arguments:
        flag_key  — the FeatBit feature flag key (from definition.md)
        variant   — variant value, e.g. "false" / "true" / "control" / "v2"
        metric    — event name, e.g. "click_start_chat"
        start     — observation window start, ISO 8601 date string
        end       — observation window end, ISO 8601 date string (or "open")

    Returns:
        (n, k) where n = number of unique users exposed to this variant
                      k = number of those users who fired the metric event

    See references/data-source-guide.md for ready-to-use implementations.
    """
    raise NotImplementedError(
        "Implement fetch_metric_summary() for your data source.\n"
        "See .featbit-release-decision/references/data-source-guide.md:\n"
        "  §FeatBit  — FeatBit experiment results API\n"
        "  §Database — SQL aggregation query\n"
        "  §Custom   — your own metrics service\n"
    )


# ── Definition parser (mirrors analyze-bayesian.py) ────────────────────────────────────

def parse_definition(path: Path) -> dict:
    text = path.read_text()
    kv = {}
    for line in text.splitlines():
        if line and not line.startswith(" ") and not line.startswith("#") and ":" in line:
            key, _, value = line.partition(":")
            kv[key.strip()] = value.strip()

    ctrl_m  = re.search(r"control:\s*(\S+)", text)
    trt_m   = re.search(r"treatment:\s*(\S+)", text)
    start_m = re.search(r"start:\s*(\S+)", text)
    end_m   = re.search(r"end:\s*(\S+)", text)

    guardrail_events: list[str] = []
    in_guardrail = False
    for line in text.splitlines():
        if "guardrail_events:" in line:
            in_guardrail = True
            continue
        if in_guardrail:
            s = line.strip()
            if s.startswith("- "):
                guardrail_events.append(s[2:].strip())
            elif s and not s.startswith("#"):
                in_guardrail = False

    return {
        "flag_key":          kv.get("flag_key", ""),
        "primary_metric":    kv.get("primary_metric_event", ""),
        "guardrail_events":  guardrail_events,
        "control_variant":   ctrl_m.group(1)  if ctrl_m  else "control",
        "treatment_variant": trt_m.group(1)   if trt_m   else "treatment",
        "start":             start_m.group(1) if start_m else "",
        "end":               end_m.group(1)   if end_m   else "open",
    }


# ── Main ──────────────────────────────────────────────────────────────────────

def main(slug: str) -> None:
    base      = Path(".featbit-release-decision") / "experiments" / slug
    defn_path = base / "definition.md"

    if not defn_path.exists():
        print(f"ERROR: {defn_path} not found.")
        sys.exit(1)

    d           = parse_definition(defn_path)
    flag_key    = d["flag_key"]
    control     = d["control_variant"]
    treatment   = d["treatment_variant"]
    start       = d["start"]
    end         = d["end"]
    all_metrics = [m for m in [d["primary_metric"]] + d["guardrail_events"] if m]

    if not flag_key:
        print("ERROR: flag_key not found in definition.md")
        sys.exit(1)
    if not all_metrics:
        print("ERROR: no metrics found in definition.md")
        sys.exit(1)

    print(f"Collecting input for: {slug}")
    print(f"  flag_key:  {flag_key}")
    print(f"  control:   {control}   treatment: {treatment}")
    print(f"  window:    {start} → {end}")
    print(f"  metrics:   {', '.join(all_metrics)}")
    print()

    result: dict = {}
    for metric in all_metrics:
        result[metric] = {}
        for variant in (control, treatment):
            print(f"  fetching {metric!r} / {variant!r} ...", end=" ", flush=True)
            n, k = fetch_metric_summary(flag_key, variant, metric, start, end)
            result[metric][variant] = {"n": n, "k": k}
            print(f"n={n}  k={k}")

    out_path = base / "input.json"
    out_path.write_text(json.dumps({"metrics": result}, indent=2) + "\n")
    print(f"\nWritten: {out_path}")
    print("Run analysis: python .featbit-release-decision/scripts/analyze-bayesian.py", slug)


if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Usage: python collect-input.py <experiment-slug>")
        sys.exit(1)
    main(sys.argv[1])

File: skills/experiment-workspace/scripts/collect-input.ts
#!/usr/bin/env npx tsx
/**
 * collect-input.ts — Query data source and write inputData to DB.
 *
 * Reads experiment definition from the DB via HTTP API,
 * calls fetchMetricSummary() for each variant × metric,
 * writes the collected inputData JSON back to the DB.
 *
 * Usage:
 *   npx tsx collect-input.ts <project-id> <experiment-slug>
 *
 * CUSTOMIZATION:
 *   Implement fetchMetricSummary() below for your data source.
 *   See references/data-source-guide.md for copy-paste patterns:
 *     §FeatBit  — call the experiment results API
 *     §Database — run a SQL aggregation query
 *     §Custom   — call your own metrics service
 *
 * Environment:
 *   SYNC_API_URL — base URL of the web app (default: http://localhost:3000)
 */

import { getExperiment, upsertExperiment, type Experiment } from "./db-client.js";

// ══════════════════════════════════════════════════════════════════════════════
// THE ONE FUNCTION YOU IMPLEMENT
// ══════════════════════════════════════════════════════════════════════════════

/**
 * Return (n_exposed, n_converted) for one variant × metric combination.
 *
 * @param flagKey   — the FeatBit feature flag key
 * @param variant   — variant value, e.g. "false" / "true" / "control" / "v2"
 * @param metric    — event name, e.g. "click_start_chat"
 * @param start     — observation window start, ISO 8601 date string
 * @param end       — observation window end, ISO 8601 date string (or "open")
 *
 * @returns { n, k } where n = users exposed, k = users who converted
 *
 * See references/data-source-guide.md for ready-to-use implementations.
 */
async function fetchMetricSummary(
  _flagKey: string,
  _variant: string,
  _metric: string,
  _start: string,
  _end: string
): Promise<{ n: number; k: number }> {
  throw new Error(
    "Implement fetchMetricSummary() for your data source.\n" +
      "See references/data-source-guide.md:\n" +
      "  §FeatBit  — FeatBit experiment results API\n" +
      "  §Database — SQL aggregation query\n" +
      "  §Custom   — your own metrics service\n"
  );
}

// ══════════════════════════════════════════════════════════════════════════════
// HELPERS
// ══════════════════════════════════════════════════════════════════════════════

function parseVariants(exp: Experiment): {
  control: string;
  treatments: string[];
} {
  const control = exp.controlVariant ?? "control";
  const treatment = exp.treatmentVariant ?? "treatment";
  const treatments = treatment.includes(",")
    ? treatment.split(",").map((s) => s.trim())
    : [treatment];
  return { control, treatments };
}

function parseGuardrails(exp: Experiment): string[] {
  if (!exp.guardrailEvents) return [];
  try {
    return JSON.parse(exp.guardrailEvents);
  } catch {
    return [];
  }
}

// ══════════════════════════════════════════════════════════════════════════════
// MAIN
// ══════════════════════════════════════════════════════════════════════════════

async function main(projectId: string, slug: string): Promise<void> {
  // 1. Read experiment from DB
  const exp = await getExperiment(projectId, slug);

  const { control, treatments } = parseVariants(exp);
  const guardrailEvents = parseGuardrails(exp);
  const primaryEvent = exp.primaryMetricEvent ?? "";
  const allMetrics = [primaryEvent, ...guardrailEvents].filter(Boolean);
  const allVariants = [control, ...treatments];

  // Use project-level flagKey if available (read from project via experiment)
  const flagKey = slug; // Default to slug; override in fetchMetricSummary if needed

  const start = exp.observationStart
    ? new Date(exp.observationStart).toISOString().split("T")[0]
    : "";
  const end = exp.observationEnd
    ? new Date(exp.observationEnd).toISOString().split("T")[0]
    : "open";

  if (!allMetrics.length) {
    console.error("ERROR: no metrics found in experiment definition");
    process.exit(1);
  }

  console.log(`Collecting input for: ${slug}`);
  console.log(`  control:   ${control}   treatment: ${treatments.join(", ")}`);
  console.log(`  window:    ${start} → ${end}`);
  console.log(`  metrics:   ${allMetrics.join(", ")}`);
  console.log();

  // 2. Collect data
  const result: Record<string, Record<string, { n: number; k: number }>> = {};
  for (const metric of allMetrics) {
    result[metric] = {};
    for (const variant of allVariants) {
      process.stdout.write(`  fetching ${metric} / ${variant} ... `);
      const { n, k } = await fetchMetricSummary(
        flagKey, variant, metric, start, end
      );
      result[metric][variant] = { n, k };
      console.log(`n=${n}  k=${k}`);
    }
  }

  // 3. Write inputData to DB
  const inputData = JSON.stringify({ metrics: result });
  await upsertExperiment(projectId, slug, { inputData });

  console.log(`\nInput data written to DB for ${projectId}/${slug}`);
  console.log(
    `Run analysis:\n  npx tsx analyze-bayesian.ts ${projectId} ${slug}`
  );
}

// ── CLI ──────────────────────────────────────────────────────────────────────

const [projectId, slug] = process.argv.slice(2);
if (!projectId || !slug) {
  console.error(
    "Usage: npx tsx collect-input.ts <project-id> <experiment-slug>"
  );
  process.exit(1);
}
main(projectId, slug);
